jena-arq/Grammar/master.jj

```diff
@@ -2176,6 +2251,19 @@ String String() : { Token t ; String lex ; }
     }
 }
 
+#ifdef ARQ
+Number Number() : { Token t ; Number number ; }
+{
+  (
+    t = < INTEGER > { number = integerValue(t.image) ; }
+  | t = < DECIMAL > { number = doubleValue(t.image) ; }
```

Comment:
Any reason to exclude doubles?

Reply:
Oh, no reason at all. Just forgot about it I think. Re-reading that method, I almost feel like removing the number parameter, and just replacing by `return integetValue(t.image);` etc. But will just add the `< DOUBLE >` for now :-)

Reply:
Would this be simpler? 

Generate `Node` from the parser, 
The data  structure is then `Map<String, Node> jsonMapping`.
To use, lookup variables in the binding to get a `Node` and also use the `Node`s from `jsonMapping`.

In the grammar, use `NumericLiteral() | BooleanLiteral()`, which return `Node`, then the sign of numbers is also handled.  as well as booleans. When needing the item for making the JSON, apply `RDFTerm2Json`, as is done when a variable is used and turned into a Json value.

There isn't a standard alone string rule returning a `Node`. The `String()` returns a string (it is used to make a RDF Literal, optional ^^ and @ parts) - to unify handling turn the string into a `Node` with `NodeFactory.createLiteral(t.image)`.

Reply:
Oh! It was by accident that it got fixed I think?

I copied parts of the `SelectClause()` to `JsonClause()`. More especifically, the Select had:

```
      // Expressions without ()
      ( LOOKAHEAD(2)
        expr = BuiltInCall()        { getQuery().addResultVar((Var)null, expr) ; }
      | expr = FunctionCall()       { getQuery().addResultVar((Var)null, expr) ; }
      | n = RDFLiteral()            { getQuery().addResultVar((Var)null, NodeValue.makeNode(n)) ; }
      | n = NumericLiteral()        { getQuery().addResultVar((Var)null, NodeValue.makeNode(n)) ; }
      | n = BooleanLiteral()        { getQuery().addResultVar((Var)null, NodeValue.makeNode(n)) ; }
```

Which became the following for the JSON counterpart:

```
    o = Var() { getQuery().addResultVar((Var)o) ; getQuery().addJsonMapping(s, o) ; }
  | o = RDFLiteral() { getQuery().addResultVar(s, NodeValue.makeNode(o)) ; getQuery().addJsonMapping(s, o) ; }
  | o = NumericLiteral() { getQuery().addResultVar(s, NodeValue.makeNode(o)) ; getQuery().addJsonMapping(s, o) ; }
  | o = BooleanLiteral() { getQuery().addResultVar(s, NodeValue.makeNode(o)) ; getQuery().addJsonMapping(s, o) ; }
```

More or less the same. But we have the Var (i.e. for cases like `"title" : ?title`), and then I did not include the expressions (hence the lookahead removed).

Does it make sense @afs ? If so, we can actually remove the `Number()` and this whole #if ARQ block. And I believe the code is doing more or less what you suggested?

Thanks!
Bruno

Reply:
Observa
tion: The JSON clause can be DRY-ed with `JsonObjectMember() ( <COMMA> JsonObjectMember()) *` and put lines 342-355 in `JsonObjectMember`.

Reply:
Couldn't reply to the number point - I think the rebase has moved things around - it now looks right.  Tests will tell!

Reply:
@afs nice catch in that observation. Reduced code duplication and avoids now possible "forgot to apply to the other part of the json clause" issues in the future :-)

Change applied. Rebased, ran all tests, tried three queries mentioned in this pull request (we have a few already! :smiley: ), got no issues.

Pushing the latest code, rebased onto the maste branch.